### PR TITLE
g4music: init at 2.4

### PR DESCRIPTION
**Pull request by magnouvean**
###### Description of changes

g4music is a small and sleek looking music player built using gtk4/libadwaita. See [here](https://gitlab.gnome.org/neithern/g4music) for more. Packaging has recently been [requested](https://github.com/NixOS/nixpkgs/issues/242405).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by magnouvean** [comment]
Any chance this could be merged? Nothing too special going on in the package and it build/works well on my machine.

File: pkgs/applications/audio/g4music/default.nix
Comment (uninsane) [inline pkgs/applications/audio/g4music/default.nix:49]:
```suggestion
    license = licenses.gpl3Only;
```

going by <https://gitlab.gnome.org/neithern/g4music/-/blob/master/COPYING> and meson.build, unless you saw the author say "gpl3 _or later_" somewhere in which case `gpl3Plus`.

**Review by uninsane — COMMENTED** [review]
built locally & running smoothly for me after applying the modifications above.

upstream moves fast & is on 3.2 now. seems to not actually require any modifications other than updating the hash to `sha256-BlHOYD4sOmJPNMzM5QA97Ah1N9tIat0Y6qxN6c5pmsw=`.

btw if you add `gobject-introspection` to `nativeBuildInputs` then this package supports cross compilation (e.g. `nix build '.#pkgsCross.aarch64-multiplatform.g4music'`). it's got a nice adaptive UI, seems like the type of thing that mobile nixos users might enjoy :)

File: pkgs/applications/audio/g4music/default.nix
Comment (magnouvean) [reply pkgs/applications/audio/g4music/default.nix]: This appears to have been needed in the previous version (2.4), but I have now moved the version to 3.2 (see latest commit) and thus I have also removed this now

**Comment by magnouvean** [comment]
Thanks for the suggestions @uninsane! I have fixed these suggestions now and it all builds well on my end as well (I have tested). I did add the `gobject-introspection` as suggested, though I'm not sure if my system is set up correctly for it or something as the provided command didn't work on my end (this was before any build happened, so not due to the recipe), so I can't confirm if the aarch64 build succeeds.

**Review by uninsane — APPROVED** [review]

File: pkgs/applications/audio/g4music/default.nix
Comment (drupol) [inline pkgs/applications/audio/g4music/default.nix:15]: I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

**Review by drupol — APPROVED** [review]
All good! Waiting for the CI to be done before merging!

File: pkgs/applications/audio/g4music/default.nix
Comment (magnouvean) [reply pkgs/applications/audio/g4music/default.nix]: I see, wasn't aware of that, but will be with future recipes :). I've now pushed an update to include this pattern instead of rec.

**Comment by uninsane** [comment]
@magnouvean have you tried playing any mp4 audio files with this (file extension is usually `.m4a`)? g4music was getting stuck on them at the 0:00 timestamp marker -- until i added `gst-plugins-bad` to `buildInputs` and now they play fine.

grepping `audio/mp4` (mime type) in gstreamer codebase hints those indeed don't have support in base or good. not really familiar with the packaging idioms around gstreamer stuff though.

**Comment by magnouvean** [comment]
> @magnouvean have you tried playing any mp4 audio files with this (file extension is usually `.m4a`)? g4music was getting stuck on them at the 0:00 timestamp marker -- until i added `gst-plugins-bad` to `buildInputs` and now they play fine.
> 
> grepping `audio/mp4` (mime type) in gstreamer codebase hints those indeed don't have support in base or good. not really familiar with the packaging idioms around gstreamer stuff though.

I did forget to test files other than mp3 @uninsane, so this doesn't suprise me. I think I'll add a seperate pr seeing as this is already merged, and I'll make sure to double check some other filetypes as well to ensure they're all good!

**Pull request merged**
